gctree/branching_processes.py

```diff
@@ -987,3 +964,253 @@ def f(x):
+        # psuedocount for the root
+        cm_list[0] = (1, 1)
+    # extract vector of function values and gradient components
+    logf_data = [CollapsedTree._ll_genotype(c, m, p, q) for c, m in cm_list]
+    logf = np.array([[x[0] * count_ls[i]] for i, x in enumerate(logf_data)]).sum()
+    grad_ll_genotype = np.array(
+        [x[1] * count_ls[i] for i, x in enumerate(logf_data)]
+    ).sum(axis=0)
+    return logf, grad_ll_genotype
+
+
+def _llforest(
+    cm_countlist,
+    p: np.float64,
+    q: np.float64,
+    marginal: bool = False,
+) -> Tuple[np.float64, np.ndarray]:
+    r"""Log likelihood of branching process parameters :math:`(p, q)` given `cm_countlist`, an iterable
+    containing tuples `(cm_counts, mult)`, where `cm_counts` is an iterable describing a tree, and `mult`
+    is the number of trees in the forest which can be described with the iterable `cm_counts`.
+
+    `cm_counts` is in the format expected as an argument to :meth:`_lltree`.
+
+    If ``marginal=False`` (the default), compute the joint log likelihood
+    .. math::
+        \ell(p, q; T, A) = \sum_{i=1}^n\log\mathbb{P}(T_i, A_i \mid p, q),
+    otherwise compute the marginal log likelihood
+    .. math::
+        \ell(p, q; T, A) = \log\left(\sum_{i=1}^n\mathbb{P}(T_i, A_i \mid p, q)\right).
+    Args:
+        p: branching probability
+        q: mutation probability
+        marginal: compute the marginal likelihood over trees, otherwise compute the joint likelihood of trees
+    Returns:
+        Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`
+    """
+    terms = [[_lltree(cmcounts, p, q), count] for cmcounts, count in cm_countlist]
+    ls = np.array([term[0][0] for term in terms])
+    grad_ls = np.array([term[0][1] for term in terms])
+    # This can be done ahead of time
+    count_ls = np.array([term[1] for term in terms])
+    if marginal:
+        # we need to find the smallest derivative component for each
+        # coordinate, then subtract off to get positive things to logsumexp
+        grad_l = []
+        for j in range(len((p, q))):
+            i_prime = grad_ls[:, j].argmin()
+            b = (grad_ls[:, j] - grad_ls[i_prime, j]) * count_ls
+            # believe it or not, logsumexp can't handle 0 in b
+            # when np.seterr(underflow='raise') on newer processors:
+            grad_l.append(
+                grad_ls[i_prime, j]
+                + np.exp(
+                    scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
+                    - scs.logsumexp(ls - ls[i_prime], b=count_ls)
+                )
+            )
+        # count_ls shouldn't have any zeros in it...
+        return (-np.log(count_ls.sum()) + scs.logsumexp(ls, b=count_ls)), np.array(
+            grad_l
+        )
+    else:
+        return (ls * count_ls).sum(), np.array(
+            [(grad_ls[:, 0] * count_ls).sum(), (grad_ls[:, 1] * count_ls).sum()]
+        )
+
+
+def fit_branching_process(
```

Comment:
I would think you could just replace the existing `CollapsedForest.mle` function with the contents of this function, instead of adding to the API